errors/error_test.go

```diff
@@ -0,0 +1,185 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package errors_test
+
+import (
+	stderrors "errors"
+	stdfmt "fmt"
+	"os"
+	"testing"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/madlambda/spells/assert"
+	"github.com/mineiros-io/terramate/errors"
+)
+
+var E = errors.E
+
+var (
+	syntaxError   errors.Kind = "syntax error"
+	tmSchemaError errors.Kind = "terramate schema error"
+	tfSchemaError errors.Kind = "terraform schema error"
+)
+
+func TestNoArgs(t *testing.T) {
+	defer func() {
+		err := recover()
+		if err == nil {
+			t.Fatal("E() did not panic")
+		}
+	}()
+	_ = E()
+}
+
+func TestErrorString(t *testing.T) {
+	type testcase struct {
+		name string
+		args []interface{}
+		want string
+	}
+
+	for _, tc := range []testcase{
+		{
+			name: "simple message",
+			args: []interface{}{
+				"error",
+			},
+			want: "error",
+		},
+		{
+			name: "simple message with kind",
+			args: []interface{}{
+				syntaxError,
+				"failed to parse config",
+			},
+			want: fmt("%s: failed to parse config", syntaxError),
+		},
+		{
+			name: "the kind of previous error is promoted if new one lacks it",
+			args: []interface{}{
+				"failed to parse config",
+				E(syntaxError, "unexpected IDENT"),
+			},
+			want: fmt("%s: failed to parse config: unexpected IDENT", syntaxError),
+		},
+		{
+			name: "multiple different error kinds",
+			args: []interface{}{
+				tmSchemaError, "failed to parse config",
+				E(tfSchemaError, "malformed terraform code"),
+			},
+			want: fmt(
+				"%s: failed to parse config: %s: malformed terraform code",
+				tmSchemaError, tfSchemaError,
+			),
+		},
+		{
+			name: "the file range gets promoted if current error lacks the file context",
+			args: []interface{}{
+				"failed to parse config",
+				E(tmSchemaError, "unexpected attribute name", hcl.Range{
+					Filename: "test.tm",
+					Start:    hcl.Pos{Line: 1, Column: 5, Byte: 3},
+					End:      hcl.Pos{Line: 1, Column: 10, Byte: 13},
+				}),
+			},
+			want: fmt("test.tm:1,5-10: %s: failed to parse config: unexpected attribute name",
+				tmSchemaError),
+		},
+		{
+			name: "nested errors",
+			args: []interface{}{
+				"1", E("2", E("3")),
+			},
+			want: "1: 2: 3",
+		},
+		{
+			name: "nested errors with last one with range",
+			args: []interface{}{
+				"1", E("2", E("3", hcl.Range{
+					Filename: "test.tm",
+					Start:    hcl.Pos{Line: 1, Column: 5, Byte: 3},
+					End:      hcl.Pos{Line: 1, Column: 10, Byte: 13},
+				})),
+			},
+			want: "test.tm:1,5-10: 1: 2: 3",
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			e := errors.E(tc.args...)
+			assert.EqualStrings(t, tc.want, e.Error())
+		})
+	}
+}
+
+func TestErrorIs(t *testing.T) {
+	type testcase struct {
+		name    string
+		args    []interface{}
```

Comment:
```suggestion
		err    E
```

Not so sure what we gained with defining test cases like this:

```go
{
			name:   "nil target is not equal",
			args:   []interface{}{"any error"},
			target: nil,
},
```

Instead of (which IMHO reads better):

```go
{
			name:   "nil target is not equal",
			args:   E("any error"),
			target: nil,
},
```

Unless we test when the constructor panics or something, but doesn't seem to be the case here.

Reply:
yep, this is leftover from the previous iteration.